src/buffer.rs

```diff
@@ -13,23 +13,44 @@ pub enum Action {
 }
 
 impl Action {
-    pub fn do_on(&self, buf: &mut Buffer) {
+    pub fn do_on(&self, buf: &mut Buffer) -> Option<usize> {
         match *self {
-            Action::Insert { start, ref text } => buf.insert_raw(start, &text[..]),
+            Action::Insert { start, ref text } => {
+                buf.insert_raw(start, &text[..]);
+                Some(start)
+            }
             Action::Remove { start, ref text } => {
-                buf.remove_raw(start, start + text.len());
+                let text_len = text.len();
+                buf.remove_raw(start, start + text_len).len();
+                if text_len > start {
+                    Some(0)
+                } else {
+                    Some(start - text.len())
+                }
             }
-            Action::StartGroup | Action::EndGroup => {}
+            Action::StartGroup | Action::EndGroup => None,
         }
     }
 
-    pub fn undo(&self, buf: &mut Buffer) {
+    pub fn undo(&self, buf: &mut Buffer) -> Option<usize> {
         match *self {
             Action::Insert { start, ref text } => {
-                buf.remove_raw(start, start + text.len());
+                buf.remove_raw(start, start + text.len()).len();
+                Some(start)
+            }
+            Action::Remove { start, ref text } => {
+                buf.insert_raw(start, &text[..]);
+                Some(start)
             }
-            Action::Remove { start, ref text } => buf.insert_raw(start, &text[..]),
-            Action::StartGroup | Action::EndGroup => {}
+            Action::StartGroup | Action::EndGroup => None,
+        }
+    }
+
+    pub fn get_start_and_text(&self) -> Option<(usize, Vec<char>)> {
```

Comment:
Note on maybe a fun future thing, replace Vec<char> with &[char] and try to get that to work.  It will save an allocation (clone) and might be an interesting lifetime exercise (will introduce a lifetime and cascade to the code that uses this...).  Might also be a PITA, not sure, but might be interesting.  Not holding up merge on this, just a thought.